### PR TITLE
fix(diagnostics): reject unknown feature names

### LINKED ISSUE #134: v5: add shared Marionette diagnostics with stable codes
## Goal
Use the existing MarionetteError taxonomy with stable rule codes for framework invariant failures.

## Dependencies
Requires the catalog and repository-local generated diagnostic contract in #129. Public website publication is deferred to post-v5 issue #177.

## Agent-development failure addressed
Incidental TypeErrors and prose-only messages cannot be classified reliably by agents, tests, lint rules, or docs.

## Public behavior contract
Preserve and formalize existing names such as `ViewError`, `CollectionViewError`, `NoChildViewError`, and `InvalidChildViewError`. Codes are stable at v5 final; messages may improve. Stable codes and repository-generated docs do not depend on a live website. Do not standardize accidental JavaScript exceptions.

## Runtime cost category
Exceptional production paths. Compact codes are allowed; the full rule catalog stays out of core.

## Acceptance criteria
- [x] Define and export the public error shape and code lookup contract.
- [ ] Seed conversion with unknown Region operations, unknown literal `@ui` references, missing literal string handlers, unknown Radio/reset operations, and other verified raw TypeError or silent-failure paths.
- [x] Preserve meaningful existing error names while giving each invariant a stable code.
- [x] Validate codes and repository-generated documentation anchors against #129 without requiring a live website.
- [x] Do not point runtime messages at unpublished `docs.marionettejs.com` routes; #177 owns public link cutover after v5 launch.
- [x] Test code stability separately from message prose.
- [x] Verify stack behavior on engines with and without `Error.captureStackTrace`.
- [x] Document SemVer and deprecation handling.

## Rollback or deprecation condition
Before stable, revise codes through the catalog. After stable, deprecate rather than reuse or silently change them.

### PR DESCRIPTION
**Pull request by paulfalgout**
Closes #134

## What

- Reject unknown or non-string `setEnabled` names with stable diagnostic code `MN0027` before mutating the feature registry.
- Preserve writes and exact return values for all documented flags, and preserve `isEnabled(unknown) === false`.
- Add the diagnostic catalog entry, compatibility guidance, release-note entry, focused coverage, and regenerated main distributions.

## Why

`setEnabled` silently created application-owned flags and allowed prototype-sensitive names such as `__proto__`. This was the final verified silent-failure path in the shared diagnostics rollout.

## Scope

This changes only explicit `setEnabled` calls. The three documented Marionette flags remain writable; `isEnabled` and internal feature reads are unchanged. Application-owned custom flags must move to application configuration or another application registry.

## Deployment Impact

No website, package, tag, release, or deployment is performed by this PR. The normal package build must include the regenerated main ESM, CommonJS, UMD, and minified artifacts.

## Testing

- `npm test -- --run test/unit/runtime/features.spec.js --coverage.enabled=false`
- `npm run check:diagnostics`
- `npm run docs:check`
- `npm run lint:ci`
- `npm run coverage` (86 files, 1,587 tests, 100% statements/branches/functions/lines)
- `npm run check:dist`
- `npm run test:source`
- `npm run test:dist`
- `npm run test:fixtures`
- `npm run size`
- `npm run performance:timing`
- `git diff --check`

Exact-base Brotli-11 comparison: +226 bytes cumulative (+0.377%); the largest individual artifact increase is +0.54%, below the 1% approval threshold.

---
## Summary by cubic
Rejects unknown or non-string `setEnabled` names with diagnostic code `MN0027`, leaving the feature registry unchanged. `setEnabled` previously created application-owned flags silently and allowed prototype-sensitive names such as `__proto__`; the registry now uses a null prototype so `isEnabled` reads are safe too.

**Migration**
- Application-owned custom flags must move to application configuration or another application registry.
- The documented Marionette flags remain writable; `isEnabled(unknown)` still returns `false`.

<sup>Written for commit fec6f79406511d8f9566919f2a8f58f4b9815039. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/marionettejs/marionette/pull/299?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by paulfalgout** [comment]
**Not approving yet** — reviewed at `453d984a` (the SHA I fetched, after the docs/spec push). One blocking finding, and it is a one-line fix.

## `setEnabled` was hardened; `isEnabled` was not

`setEnabled` now rejects prototype-sensitive names via `Object.hasOwn`. Its sibling reader still does a raw bracket read:

```js
function isEnabled(name) {
  return !!FEATURES[name];
}
```

`FEATURES` is a plain object literal, so that read walks the prototype chain. Against the real module at `453d984a`:

```
"__proto__"        setEnabled -> MN0027    | isEnabled -> true
"constructor"      setEnabled -> MN0027    | isEnabled -> true
"toString"         setEnabled -> MN0027    | isEnabled -> true
"hasOwnProperty"   setEnabled -> MN0027    | isEnabled -> true
"nope"             setEnabled -> MN0027    | isEnabled -> false
```

The writer rejects these names loudly while the reader silently answers `true` for them.

This contradicts two contracts the PR itself asserts:

1. The PR body: *"preserve `isEnabled(unknown) === false`"*. It does not hold for the four names above.
2. [`docs/features.md:24`](docs/features.md:24) documents `isEnabled('fooFlag'); // false` — the stated contract is unknown ⇒ `false`.

It also inverts the PR's stated purpose. The description calls this *"the final verified silent-failure path"*, but the silent half is the reader. After this PR, `setEnabled('__proto__')` throws loudly — an improvement — while `isEnabled('__proto__')` still silently returns `true`. The loud half got fixed and the quiet one was left in place.

Scope is bounded and worth stating plainly: every internal read passes a literal documented name, so no shipped Marionette behavior changes. The exposure is application code that forwards a config- or user-derived string into the public `isEnabled`.

## Why the suite stays green

The hazard names are in the spec, but only under `#setEnabled`. The `#isEnabled` block covers unknown names with `'applicationOwned'` alone, and non-string names with `undefined, null, 0, {}, Symbol` — all of which correctly yield `false`. No assertion ever puts a prototype key through `isEnabled`. Their presence in the file makes the gap read as covered on a skim.

## Suggested fix

Own-key discipline in the reader, mirroring the writer. Not a throw — `isEnabled` is a predicate and should stay total:

```js
function isEnabled(name) {
  return Object.hasOwn(FEATURES, name) && !!FEATURES[name];
}
```

`Object.hasOwn` is already load-bearing in `mixins/events.js` and `utils/assign-in.js`, so this introduces no new runtime floor. It short-circuits to `false` for non-string and non-own names, and preserves the existing `!!` coercion for documented flags, so the current `#isEnabled` assertions keep passing unchanged. This is a runtime change, so `dist/` needs regenerating.

Worth extending the reader block with the same name list the writer block uses, so the two stay locked together.

## What is already correct

- `Object.hasOwn` rather than `in` or a truthiness test — the right primitive, and it is what makes `__proto__` rejection real rather than incidental.
- Non-string check ordered first, so `Object.hasOwn` never receives a coercible key.
- `return FEATURES[name] = state` untouched — the exact-state return is preserved, and the spec pins it with an object identity check.
- The newest push tightened the writer further with `''`, `' '`, and a whole-registry mutation check.
- MN0027 catalog entry carries a slug and actionable remediation.
- CI green at this head; no failing checks.

The hardening approach is right. It is applied to one of the two functions that need it.

**Comment by paulfalgout** [comment]
Reviewed with Claude Code at exact head `453d984`: **APPROVED — no blocking code issues found.**

Claude reviewed the final `setEnabled` guard, `MN0027` contract, compatibility disclosure, atomicity coverage, read/write asymmetry, regenerated distributions, internal-call-site audit, and exact-base size evidence. The earlier conditional pass caused the changelog entry, rejected-name context, non-string read coverage, empty/whitespace cases, registry snapshots, and concrete v4-to-v5 custom-flag migration example now present at this head.

The remaining observations are non-blocking: this intentionally ships only in the v5 major, state values remain unvalidated by the existing contract, and future internal callers are still protected by the runtime guard plus ordinary test coverage rather than a dedicated source-scan rule.

<sub>Reviewed with Claude Code.</sub>

**Comment by paulfalgout** [comment]
Reviewed with Claude Code at exact head `6888a5ab0081d68591b8bee795157217c62eb47d`: **APPROVED — no blocking code issues found.**

The prior prototype-key reader blocker is resolved without adding hot-path work: the static `FEATURES` object now uses the object-literal `__proto__: null` form, while `isEnabled` remains unchanged. The shared invalid-name matrix proves `constructor`, `toString`, `__proto__`, `hasOwnProperty`, empty, whitespace, and ordinary unknown names all read as `false` and are rejected atomically by `setEnabled` with `MN0027`.

Claude confirmed the null-prototype literal is the minimal fix and that the earlier blocker is resolved. A repository search found no direct inherited-method use on `FEATURES`; regenerated ESM, CommonJS, UMD, and minified artifacts preserve the null-prototype literal. Full coverage, source/dist validation, lint, diagnostics, size, timing, and post-commit `check:dist` pass at this head.

<sub>Reviewed with Claude Code.</sub>

File: test/unit/runtime/features.spec.js [hunk 1]
```diff
@@ -1,17 +1,83 @@
...
-  it('enabled when its present and true', function() {
-    setEnabled('foo', true);
-    expect(isEnabled('foo')).to.be.true;
-  });
+  describe('#isEnabled', function() {
+    it('returns the configured state of each Marionette feature', function() {
+      for (const name of Object.keys(FEATURES)) {
+        const initialState = FEATURES[name];
+
+        setEnabled(name, !initialState);
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply test/unit/runtime/features.spec.js:25]: Fixed in 1aabff3. A shared afterEach restores every original feature value and deletes any stray key, so cleanup runs even when an assertion fails.

File: test/unit/runtime/features.spec.js [hunk 2]
```diff
@@ -1,17 +1,83 @@
...
   });
 
-  it('disabled when not present', function() {
-    expect(isEnabled('foo')).to.be.false;
+  describe('#setEnabled', function() {
+    it('sets and returns the exact state for a known feature', function() {
+      const state = {};
+
+      expect(setEnabled('childViewEventPrefix', state)).to.equal(state);
+      expect(FEATURES.childViewEventPrefix).to.equal(state);
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply test/unit/runtime/features.spec.js]: Fixed in 1aabff3. The same shared afterEach restores childViewEventPrefix and all other original feature values even when an assertion fails.

File: runtime/features.js [hunk 3]
```diff
@@ -11,6 +14,20 @@ function isEnabled(name) {
...
 
 function setEnabled(name, state) {
+  if (typeof name !== 'string') {
+    throw new MarionetteError({
+      code: 'MN0027',
+      message: 'The feature name must be a documented Marionette feature name.'
+    });
+  }
+
+  if (!Object.hasOwn(FEATURES, name)) {
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply runtime/features.js]: Fixed in 6888a5a without adding hot-path work. FEATURES now has a null prototype, so the unchanged predicate returns false for constructor, toString, __proto__, hasOwnProperty, and other unknown names; the shared regression matrix covers each case.

**Comment by paulfalgout** [comment]
Reviewed with Claude Code at exact head `1aabff359a8aca5056cb34fe1a4ab65691a56edc`: **APPROVED — no blocking code issues found.**

This head is test-only relative to the previously approved runtime head. It adds shared `afterEach` cleanup that restores the singleton feature values and deletes any unexpected key even when an assertion fails, directly addressing both valid Cubic isolation findings. Runtime source and generated distributions remain identical to approved head `6888a5a`.

Focused tests, lint, the full 1,587-test suite, 100% coverage, and `git diff --check` pass. Claude noted only a non-blocking future-harness edge case for tests that might directly delete an original registry key; no current test does so, and production code does not expose deletion.

<sub>Reviewed with Claude Code.</sub>

File: test/unit/runtime/features.spec.js [hunk 4]
```diff
@@ -1,17 +1,91 @@
...
-    expect(isEnabled('foo')).to.be.true;
+  afterEach(function() {
+    for (const name of Object.keys(FEATURES)) {
+      if (!Object.hasOwn(initialFeatures, name)) {
+        delete FEATURES[name];
+        continue;
+      }
+
+      setEnabled(name, initialFeatures[name]);
+    }
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply test/unit/runtime/features.spec.js:17]: Fixed in fec6f79. Cleanup now deletes unexpected current keys, then iterates every original key and restores it directly. Direct assignment is intentional here because setEnabled correctly rejects a documented key after a test has deleted it.

**Comment by paulfalgout** [comment]
Reviewed with Claude Code at exact head `fec6f79406511d8f9566919f2a8f58f4b9815039`: **APPROVED — no blocking code issues found.**

This head is test-only relative to the approved runtime head. The final cleanup fix restores test isolation in two phases: remove unexpected keys, then directly restore every original registry key, including one a failing test may have deleted. Direct assignment is intentional because the public `setEnabled` guard correctly rejects a documented key after deletion.

Focused tests, lint, the full 1,587-test suite, 100% coverage, and `git diff --check` pass. Runtime and generated distributions remain identical to approved head `6888a5a`.

<sub>Reviewed with Claude Code.</sub>
